### DISCUSSION TITLE
craft/web/ErrorHandler::renderException re-add left attributes from Yii2 or add an event to add those

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#11799] [Ideas] [answered]
I usually control my REST Application states via Exceptions. In Craft 3 [Yii2 did everything for me.](https://github.com/yiisoft/yii2/blob/master/framework/web/ErrorHandler.php#L150-L157) 
When I throw certain Exceptions, I was able to tell the user the exact reason why something has failed in JS. [Craft 4 unfortunately removed that option](https://github.com/craftcms/cms/blob/develop/src/web/ErrorHandler.php#L134-L138) unless I overwrite your ErrorHandler with the one from Yii2 again.

For example in my Controller
```PHP
$i = $this->request->getBodyParam('someParam');
if($i === 0){
    throw new HttpException(400, 'i must not be 0', 1);
}

if($i > 10){
  throw new HttpException(400, 'i must not be larger than 10', 2);
}
```

the responses would then be like
```JS
{
    name:  'Bad Request',
    message: 'i must not be 0' | 'i must not be larger than 10',
    status: 400,
    code: 0 | 1
}
```
I was able to check 
```JS
// display the error message
if (res.code === 0) {
   // do something special, like prepare a new request with the better value instantly
}
// and so on
``` 

However: Craft 4 does only this in a live environment
```JS
{
    name:  'Bad Request',
    message: 'Bad Request'
}
```
This one doesn't contain any useful information at all and a totally different response in dev mode.

It would be awesome if there was an event for us so I can easily overwrite your response with the one from Yii2 again without the requirement to overwrite the entire class again. In a perfect world you would just call Yii2 `convertExceptionToArray` to have all relevant information in all application states.

Even more problematic now: depending of the response header, totally different data is passed... In Yii2 the structural exception response was always the same. Craft 4 shows a detailed response in case of an XML request and basically nothing in case of JSON. (To avoid confusions: I frequently use https://www.yiiframework.com/doc/api/2.0/yii-filters-contentnegotiator and it's a requirement for me provide different response types for different API users)

**Comment by brandonkelly** [comment, marked as answer]
This actually changed back in Craft 3.6 (5775433d21709e412434be21023a75842fe2d655).

I’ve just updated the code for the next Craft 3 and 4 releases, so all JSON error responses start including `name` and `code` keys, and `status` for `HttpException`s regardless of Dev Mode, just like Yii does out of the box. (48b7ead20e27929b20ee863ea3872ea14d5e3b04)

**Reply by Anubarak** [reply]
Yeah you are right - I didn't use the latest Craft 3 versions because I always need to inspect each commit due to these reasons 😅 
However this time around we skipped 3.6.15 directly to Craft 4 in order to safe some time.

As always I highly appreciate your incredible fast reaction. Your support is always on top 👍

**Reply by brandonkelly** [reply]
Craft 3.7.52 and 4.2.2 are out now with that change.
